src/rt/atomic.rs

```diff
@@ -144,24 +144,30 @@ pub(crate) fn fence(order: Ordering) {
 }
 
 impl State {
-    pub(super) fn last_dependent_accesses<'a>(
-        &'a self,
-        action: Action,
-    ) -> Box<dyn Iterator<Item = &'a Access> + 'a> {
-        match action {
-            Action::Load => Box::new({ self.last_load.iter().chain(self.last_store.iter()) }),
-            Action::Store => Box::new({ self.last_load.iter().chain(self.last_store.iter()) }),
-            Action::Rmw => Box::new({ self.last_load.iter().chain(self.last_store.iter()) }),
-        }
+    pub(super) fn for_each_last_dependent_access(
```

Comment:
I think that rather than changing this method to a `for_each`, it could just return an `impl Iterator` instead of a box (since all the match arms return an iterator of the same type) --- this would have the same perf benefit of not allocating, but the API would be more flexible and idiomatic.

Reply:
Methods for concrete objects can be made to return `Impl Iterator` (and although chaining does incur some cost, it is very slight). But the clients don't use them - they use the method of `Store` and it still needs to be `for_each_...` so that's not much of an improvement in my opinion. Also, as we have here a kind of polymorphism by hand, I think it makes sense for all methods to have the same structure. I'll defer to your judgment here - if you think that the `Impl Iterator` version is better, I'll implement it.

Reply:
> But the clients don't use them - they use the method of `Store` and it still needs to be `for_each_...` so that's not much of an improvement in my opinion. 

Ah, yeah — I overlooked that when I left this comment. Since the `Store` method cannot return an iterator, I think this is fine. A comment on these methods might be helpful, though?